Basic Primitive Components and code refactors

Koushith opened the pull request:
### Description:
This PR contains basic Primitive Components.

### **Added**

The following reusable components and config are added to this PR

- [x] Theming Config
- [x]  color palette file
- [x] Typography Component
- [x] Typography
- [x] Buttons
- [x] Alerts
- [x] Avatar    
- [x] Card
- [x] Quick Preview and usage example on http://localhost:3000/preview 

### **Removed**

- [x] Row and Col as we are using the external package.

### **Changed**

- Converted all JS files to TypeScript

### **Pending**

- [ ] Input
- [ ] Image

dsouzaedison committed:
Typescript implementation on Row - Example

Koushith committed:
adds basic primitive components and theming configurations

Koushith committed:
refactored code to avoid redundancy

Koushith committed:
adds basic primivtive components

Koushith committed:
adds all primitive components

Koushith committed:
modifies button component and changed the files to TypeScript

Koushith committed:
adds index file to primitive component directory

dsouzaedison commented:
1. Use proper file naming
    - Small case hyphen separated names must be used (Ex: user-card)
    - Every component should have minimum 2 files
        - user-card.component.tsx
        - user-card.component.props.ts
    - Every screen must be enclosed in a folder and should contain at least 2 files
        - user-card.screen.tsx
        - user-card.screen.props.ts
    - Other files can be named like this
        - user-card.screen.styles.ts
        - user-card.screen.state.ts (For class components)
2. No default exports. Only use named exports to ensure that IDE intellisense recognises your imports as you type.
3. Rename “primitiveComponents” folder to “primitive”
4. There is no need for template literal syntax unless you have a hard-coded string
5. Specify return types for all functions. Couldn’t find it in button utils.
6. Do not segregate styled component implementation based on CSS properties. Do it based on variants instead. At many places, separate functions are used to identify value for one CSS property such as color. Ideally, most of the functions that start with “handle” may not be necessary.
7. You may not need to define enums for CSS values. Ex: Alignments
8. Palette files should contain actual color names as property names. Ex: blue: ‘#00f’
9. Colors file can contain mapping of palette colours to context based properties. Ex: primary: palette.blue. Also need not define an interface for it. You can auto-generate it.
10. Create appropriate layout components even If any external library is used. Do not import directly from library into screens.

Koushith committed:
reorganises all folder structures, removes all default exports and other code-refactors

Koushith committed:
fixes icon component

Koushith commented:
## Recent commit adds following changes

## Adds 

- [x] Icon Component.
- [x] Image Component.
- [x] Input Component.  

## Changes

- [x] Proper naming conventions with small case and hyphen separated.
- [x] Organized folders for different screens.
- [x] Removed all default exports.
- [x] Renamed to Primitives.
- [x]  Adds return types for functions.
- [x] eliminated unnecessary enums in alignments.
 

## WIP 
- [ ] Color palette values swapping.
- [ ]  segregation of styled component implementation based on CSS properties.

dsouzaedison commented on src/components/primitive/Icons/icon.component.props.ts:
```diff
@@ -0,0 +1,11 @@
+  size?: number;
+  color?: string;
+
+  default?: any;
```
Don't use default as a prop name since it's a keyword.

Koushith replied:
removed this, will not use it in the future.

dsouzaedison commented on src/components/primitive/Icons/icon.component.props.ts:
```diff
@@ -0,0 +1,11 @@
+interface IconProps extends React.ImgHTMLAttributes<HTMLImageElement> {
+  icon: IconType;
+  size?: number;
+  color?: string;
```
Allow colors that are there in the colors file and not any string.

Koushith replied:
replaced colors- string with ColorType.

dsouzaedison commented on src/components/primitive/Icons/icon.component.tsx:
```diff
@@ -0,0 +1,16 @@
+import { IconProps } from './icon.component.props';
+import { ReactSVG } from 'react-svg';
+import { ICONS } from './icons';
```
Avoid direct imports of static resources. Use theme hook instead.

dsouzaedison commented on src/components/primitive/Icons/icon.component.tsx:
```diff
@@ -0,0 +1,16 @@
+import { ICONS } from './icons';
+import { SvgWrapper } from './icon.component.style';
+
+export const Icon = ({ icon, className, size = 14, color, ...rest }: IconProps) => (
```
If there is a possibility that we might be using open source icon libraries where we don't have to deal with SVGs, we might need to name the component as Icon itself. In such a case, it's better to rename this component as SvgIcon.

dsouzaedison commented on src/components/primitive/Icons/icons.ts:
```diff
@@ -0,0 +1,10 @@
+import { arrowRight, arrowLeft, arrowDown, hamburgerMenu } from '../../../assets';
```
No direct imports. Use theme instead. Add an 'icons' property whose value can be accessed based on the key passed as a prop.

dsouzaedison commented on src/components/primitive/Image/image.component.props.ts:
```diff
@@ -0,0 +1,10 @@
+interface ImageComponentProps {
```
This must extend React native's ImageComponentProps.

dsouzaedison commented on src/components/primitive/Image/image.component.props.ts:
```diff
@@ -0,0 +1,10 @@
+interface ImageComponentProps {
+  id?: string | number | any;
```
Why do we need this?

dsouzaedison commented on src/components/primitive/Layouts/Col/col.component.props.ts:
```diff
@@ -0,0 +1 @@
+export {};
```
Don't need these files where nothing is exported. You can delete them.

dsouzaedison commented on src/components/primitive/Text/text.component.utils.ts:
```diff
@@ -1,56 +1,48 @@
 const handleColor = (variant: string, theme: any) => {
   switch (variant) {
-    case Variants.PRIMARY:
+    case 'primary':
```
Why are these things hard-coded? The spec was to remove such functions and organize them based on variants right?

Koushith replied:
removed all old implementation and organized according to Variants.

dsouzaedison commented on src/components/primitive/index.ts:
```diff
@@ -0,0 +1,9 @@
+export * from './Icons/icon.component';
+export * from './Text/text.component';
+export * from './Image/image.component';
+export * from './Text/text.component';
```
This is already exported above

Koushith replied:
removed duplicate export.

dsouzaedison commented on src/components/primitive/Text/text.component.styles.ts:
```diff
@@ -0,0 +1,11 @@
+import { handleColor, handleSize, handleTextAlign } from './text.component.utils';
+import { StyledTextProps } from './text.component.props';
+
+export const Typography = styled.h1<StyledTextProps>`
```
Rename it to StyledText or so and maintain consistency in naming. Typography and Text are 2 different names and could be confusing.

Koushith committed:
fixes typography component

Koushith committed:
modifies theming file

dsouzaedison commented on src/components/primitive/Typography/typography.component.tsx:
```diff
@@ -0,0 +1,48 @@
+import { Base, Title, ContentHeader, Small, Content, Description } from './typography.component.styles';
+import { TextPresets, TextProps } from './typography.component.props';
+
+const Text: React.FC<TextProps> = ({ size, color, lineHeight, upperCase, variant, children }) => {
```
File name and component name should match. Either use Text or Typography.

dsouzaedison commented on src/components/primitive/Typography/typography.component.tsx:
```diff
@@ -0,0 +1,48 @@
+  const Typography = styled(TextComponent)`
+    /* custom styles goes here */
+
+    ${({ center }) => (center ? `text-align: center` : '')};
```
Do we have access to component props here?

dsouzaedison commented on src/components/primitive/Typography/typography.component.tsx:
```diff
@@ -0,0 +1,48 @@
+
+    ${({ upperCase }) => (upperCase ? `text-transform: uppercase` : '')};
+    ${({ lowerCase }) => (lowerCase ? `text-transform: lowercase` : '')};
+    ${({ paddingLeft }) => (paddingLeft ? `padding-left: ${paddingLeft}px` : '')};
```
Should we go with px? How would affect the responsiveness?

Koushith replied:
replaced this with a rem unit.

dsouzaedison commented on src/components/primitive/preview.tsx:
```diff
@@ -33,41 +34,13 @@ export const HeadingContainer = styled.div`
 
 export const Preview: React.FC = () => {
+  const [text, setText] = useState('');
+  console.log('text0', text);
```
No logs in PR's please.

Koushith replied:
yes, cleared all logs after testing.

dsouzaedison commented on src/themes/palette.ts:
```diff
@@ -1,67 +1,114 @@
-  buttonErrorHover: '#b54248',
+export const palette = {
+  green: {
+    greenDefault: '#2ba745',
```
Its not necessary to rewrite the color name since these properties are already contained inside that color. Example:

green: {
  main: '#0f0',
  light: '#123',
  lighter: '#123',
  dark: '#123',
  darker: '#123',
  translucent: '#123',
  contrastText: '#123',
}

dsouzaedison replied:
Please create an interface as well to maintain consistency.

dsouzaedison commented on src/themes/palette.ts:
```diff
@@ -1,67 +1,114 @@
+    greyText: '#41464b', // text on grey bg
+    greyBackground: '#8E90A6', //secondary btn color
+  },
+  neutrals: {
```
This introduces inconsistency. Only color names can appear on this object. Don't use terms like 'ghost'. It's not obvious what's the shade. Follow the above format. Keep the property names consistent.

Koushith committed:
fixes typography component

Koushith committed:
fixes Rol,Col wrapper component

Koushith committed:
fixes Rol,Col wrapper component

Koushith commented:
## Recent commit adds following changes

- [x] Wrapper Component for Row and Col.

## Changes 

- [x] Fixed naming and custom styling options in the Text component.
- [x] Renamed Typography to Text.
- [x] Removed old Typography implementation and organized based on variants.
- [x] Deletes unused files in the Primitive component directory.  

## WIP - 

-[ ]  Icons- we have to decide on any open-source Icon library.
